### ISSUE TITLE
Subscription: Implement working/secure App Store receipt validation logic on the server

### ISSUE THREAD

**Issue by danieldaquino** [damus-io/damus#1812]
## Acceptance criteria

- [ ] Ensure we can securely verify with Apple that a particular user actually paid for their susbcription before giving them access to the API.
- [ ] We should make it work on both environments: Sandbox and production

## Resources

- [App Store Server API (more recent)](https://developer.apple.com/documentation/appstoreserverapi)
- [Validating receipts with App Store (might be old)](https://developer.apple.com/documentation/storekit/in-app_purchase/original_api_for_in-app_purchase/validating_receipts_with_the_app_store)

**Comment by danieldaquino** [comment]
@jb55, @alltheseas, I have been working on this one, and I believe we are very close to finishing this ticket. I made the parts that send, parse, and verify the receipt (and those steps seem to be working on the Sandbox). The only remaining portion seems to be decoding some transaction data from some binary data and getting the expiry date for the expiry bumping.

I will keep you guys updated

**Comment by danieldaquino** [comment]
@jb55, @alltheseas, I have the initial code entirely (or mostly) working and well-formatted. I am trying to test to ensure things are working as expected, but the Sandbox is quite janky and inconsistent when it comes to erasing the purchase history to reset testing. 

Almost as if there are different caches around iOS and the App Store server that become inconsistent for a period of time after resetting the purchase history.

**Comment by danieldaquino** [comment]
@jb55, @alltheseas, sent the patches!

- Server changes: https://groups.google.com/a/damus.io/g/patches/c/0oa426ogwe4
- iOS changes: https://groups.google.com/a/damus.io/g/patches/c/lUYXJqk6afk

I was limited in terms of testing edge cases and auto-renewals because the Sandbox seemed to stop working after a certain point. Upon clearing purchase history some parts of iOS would behave as if the clearing was successful, whereas other parts of iOS behave as if there was already a purchase, causing issues to the test.

I researched online and found some other people mentioning that Sandbox behavior is often inconsistent and comes back to normal after a few hours or days. So I opened a second ticket (https://github.com/damus-io/damus/issues/1992) to remind us to further test edge cases — hopefully this issue goes away on its own.
